lib/AST/ASTScopeCreation.cpp

```diff
@@ -617,7 +617,7 @@ class ScopeCreator final {
 #endif
     ASTScopeAssert(startOrder * endOrder != -1,
                    "Start order contradicts end order");
-    return startOrder + endOrder < 1;
+    return startOrder <= 0 && endOrder <= 0;
```

Comment:
After looking at this again, I'm suspecting that the right answer is an assertion that `endOrder != 0` and a simple test for `endOrder < 0`. But I haven't tried it yet. I'll run some tests this weekend. (I would also keep the existing assertion.)

Reply:
@compnerd I've got a slightly different solution that I like better, or I will if it passes the tests. See https://github.com/apple/swift/pull/28039
Once it lands, I'm hopeful it will solve this problem Thanks for your patience!

Reply:
Thanks for fixing this, I'll test it out locally.

Reply:
Happy to do so. Thank you for your patience.